### PR TITLE
wallet rpc: return final tx hex from walletprocesspsbt if complete

### PR DESCRIPTION
**Pull request by pinheadmz**
See https://github.com/bitcoin/bitcoin/pull/28363#discussion_r1315753887

`walletprocesspsbt` currently returns a base64-encoded PSBT and a boolean indicating if the tx is "complete". If it is complete, the base64 PSBT can be finalized with `finalizepsbt` which returns the hex-encoded transaction suitable for `sendrawtransaction`.

With this patch, `walletprocesspsbt` return object will ALSO include the broadcast-able hex string if the tx is already final. This saves users the extra step of calling `finalizepsbt` assuming they have already inspected and approve the transaction from earlier steps.

### CODE REVIEW

**Review by furszy — COMMENTED** [review]
Concept ACK

Should mention the API changes in the release notes too.

**Review by ismaelsadeeq — COMMENTED** [review]
Tested ACK e3d484b603abff69c6ebfca5cfb78cf82743d090

I've create a psbt on regtest and process it with `walletprocesspbst` the output now has the hex encoded network tx, which I `sendrawtransaction` to the node and was accepted.

I think this is a nice automation, you dont have to call `finalizepsbt` if the processed psbt is complete.

File: test/functional/rpc_psbt.py [hunk 1]
```diff
@@ -217,12 +217,21 @@ def run_test(self):
...
+
+        # Alternative method: sign AND finalize in one command
+        processed_finalized_psbt = self.nodes[0].walletprocesspsbt(psbt=psbtx, finalize=True)
+        finalized_psbt = processed_finalized_psbt['psbt']
+        finalized_psbt_hex = processed_finalized_psbt['hex']
+        assert signed_psbt != finalized_psbt
+        assert finalized_psbt_hex == finalized_hex
 
         # Manually selected inputs can be locked:
         assert_equal(len(self.nodes[0].listlockunspent()), 0)
```
Comment (ismaelsadeeq) [inline test/functional/rpc_psbt.py:237]: With this PR `finalizepsbt` rpc calls to get hex encoded tx after a complete `walletprocesspsbt` in the functional tests are not necessary

I've done that in this commit https://github.com/bitcoin/bitcoin/commit/5b93ed895e5086b35f70c569e9c8f087a21e6d15
feel free to pick it in this PR.

Or I will follow up after this

**Review by BrandonOdiwuor — COMMENTED** [review]
Tested ACK [#e3d484b](https://github.com/bitcoin/bitcoin/pull/28414/commits/e3d484b603abff69c6ebfca5cfb78cf82743d090)

The PR will help reduce the steps in creating PSBT transactions by finalizing transactions with `walletprocesspsbt` if all signatures are already present

would help improve #28363  - [(check discussion)](https://github.com/bitcoin/bitcoin/pull/28363#discussion_r1315753887)

```sh
$ ./src/bitcoin-cli -signet walletcreatefundedpsbt '[]' '[{"tb1qteq3qqhsxhzle98t6k3ecpluaapg84an65rw7l": 0.0008}]'
  
{                                                                                                                                                                                                                  
  "psbt": "cHNidP8BAHECAAAAAe+Y/3akTBUt92Mw40C6GvfDGmUpJsClkO2U9IIh4T+CAQAAAAD9////Aj4EDgAAAAAAFgAUgM1wLZlob/pSB96COtrlAxtemAWAOAEAAAAAABYAFF5BEALwNcX8lOvVo5wH/O9Cg9ezAAAAAAABAHECAAAAAduDPeZZtQlruL4IoNEoQbkaS5qT
7qHcrwVSRqIT1RyAAAAAAAD+////AvMVewEAAAAAFgAUk0uFiw1qYDauGlJX8sCWb9/ZW7RAQg8AAAAAABYAFHNQXZSO8IfJS9hVety+234E0SaEy24CAAEBH0BCDwAAAAAAFgAUc1BdlI7wh8lL2FV63L7bfgTRJoQiBgMoObTjYDiO/v1NA6150z8m1FmklkxZzDvwd4TVUztXbBi
yXiGAVAAAgAEAAIAAAACAAAAAAAAAAAAAIgIDo2i8r5PEOyf1Ev0+AFJBQHyCiIcwBO1k7rHDyuaoE/8Ysl4hgFQAAIABAACAAAAAgAEAAAAAAAAAAAA=",                                                                                            
  "fee": 0.00001410,                                                                                                                                                                                               
  "changepos": 0                                                                                
} 
```

```sh
$ ./src/bitcoin-cli -signet walletprocesspsbt cHNidP8BAHECAAAAAe+Y/3akTBUt92Mw40C6GvfDGmUpJsClkO2U9IIh4T+CAQAAAAD9////Aj4E
DgAAAAAAFgAUgM1wLZlob/pSB96COtrlAxtemAWAOAEAAAAAABYAFF5BEALwNcX8lOvVo5wH/O9Cg9ezAAAAAAABAHECAAAAAduDPeZZtQlruL4IoNEoQbkaS5qT7qHcrwVSRqIT1RyAAAAAAAD+////AvMVewEAAAAAFgAUk0uFiw1qYDauGlJX8sCWb9/ZW7RAQg8AAAAAABYAFHN
QXZSO8IfJS9hVety+234E0SaEy24CAAEBH0BCDwAAAAAAFgAUc1BdlI7wh8lL2FV63L7bfgTRJoQiBgMoObTjYDiO/v1NA6150z8m1FmklkxZzDvwd4TVUztXbBiyXiGAVAAAgAEAAIAAAACAAAAAAAAAAAAAIgIDo2i8r5PEOyf1Ev0+AFJBQHyCiIcwBO1k7rHDyuaoE/8Ysl4hgF
QAAIABAACAAAAAgAEAAAAAAAAAAAA=

{
  "psbt": "cHNidP8BAHECAAAAAe+Y/3akTBUt92Mw40C6GvfDGmUpJsClkO2U9IIh4T+CAQAAAAD9////Aj4EDgAAAAAAFgAUgM1wLZlob/pSB96COtrlAxtemAWAOAEAAAAAABYAFF5BEALwNcX8lOvVo5wH/O9Cg9ezAAAAAAABAHECAAAAAduDPeZZtQlruL4IoNEoQbkaS5qT
7qHcrwVSRqIT1RyAAAAAAAD+////AvMVewEAAAAAFgAUk0uFiw1qYDauGlJX8sCWb9/ZW7RAQg8AAAAAABYAFHNQXZSO8IfJS9hVety+234E0SaEy24CAAEBH0BCDwAAAAAAFgAUc1BdlI7wh8lL2FV63L7bfgTRJoQBCGsCRzBEAiAUUXjmuYYiPzB9gLi8DWam9RL6iJ+wG5Mv0+O
UeWJPgwIgANOdu3NpxlPzOQ/oC0nkV6IRQ2rCps8x63X+u/Ta0+gBIQMoObTjYDiO/v1NA6150z8m1FmklkxZzDvwd4TVUztXbAAiAgOjaLyvk8Q7J/US/T4AUkFAfIKIhzAE7WTuscPK5qgT/xiyXiGAVAAAgAEAAIAAAACAAQAAAAAAAAAAAA==",
  "complete": true,
  "hex": "02000000000101ef98ff76a44c152df76330e340ba1af7c31a652926c0a590ed94f48221e13f820100000000fdffffff023e040e000000000016001480cd702d99686ffa5207de823adae5031b5e980580380100000000001600145e411002f035c5fc94e
bd5a39c07fcef4283d7b3024730440220145178e6b986223f307d80b8bc0d66a6f512fa889fb01b932fd3e39479624f83022000d39dbb7369c653f3390fe80b49e457a211436ac2a6cf31eb75febbf4dad3e80121032839b4e360388efefd4d03ad79d33f26d459a496
4c59cc3bf07784d5533b576c00000000"
}
```

```sh
$ ./src/bitcoin-cli -signet sendrawtransaction 02000000000101ef98ff76a44c152df76330e340ba1af7c31a652926c0a590ed94f48221e13
f820100000000fdffffff023e040e000000000016001480cd702d99686ffa5207de823adae5031b5e980580380100000000001600145e411002f035c5fc94ebd5a39c07fcef4283d7b3024730440220145178e6b986223f307d80b8bc0d66a6f512fa889fb01b932fd3
e39479624f83022000d39dbb7369c653f3390fe80b49e457a211436ac2a6cf31eb75febbf4dad3e80121032839b4e360388efefd4d03ad79d33f26d459a4964c59cc3bf07784d5533b576c00000000

176af5a2291ce02add13ec854c817389638d3883427e47ea0e17ca28be177122
```

File: test/functional/rpc_psbt.py (hunk 1 above)
Comment (pinheadmz) [reply test/functional/rpc_psbt.py:237]: fantastic thanks! I'll cherry-pick

**Comment by ismaelsadeeq** [comment]
re ACK 2e249b922762f19d6ae61edaad062f31bc2849f3

**Comment by BrandonOdiwuor** [comment]
re ACK 2e249b9

**Review by ishaanam — COMMENTED** [review]
ACK 2e249b922762f19d6ae61edaad062f31bc2849f3

It would be useful if this was also implemented for `descriptorprocesspsbt`.

**Comment by Randy808** [comment]
Tested ACK e3d484b603abff69c6ebfca5cfb78cf82743d090

**Comment by achow101** [comment]
ACK 2e249b922762f19d6ae61edaad062f31bc2849f3

**Comment by ismaelsadeeq** [comment]
> ACK [2e249b9](https://github.com/bitcoin/bitcoin/commit/2e249b922762f19d6ae61edaad062f31bc2849f3)
> 
> It would be useful if this was also implemented for `descriptorprocesspsbt`.

I think it would be useful for `descriptorprocesspsbt` also, opened PR for that @ishaanam @pinheadmz @BrandonOdiwuor 
[28492](https://github.com/bitcoin/bitcoin/pull/28492)

**Pull request merged**